src/test/java/org/jenkinsci/plugins/workflow/job/WorkflowJobTest.java

```diff
@@ -157,4 +164,38 @@ public void newBuildsShouldNotLoadOld() throws Throwable {
         });
     }
 
+    @Test
+    public void deletionShouldWaitForBuildsToComplete() throws Throwable {
+        var p = j.createProject(WorkflowJob.class);
+        p.setDefinition(new CpsFlowDefinition(
+                    "import java.time.Duration\n" +
+                    "import java.time.Instant\n" +
+                    "@NonCPS def tightLoop() {\n" +
```

Comment:
Is this really necessary? Did you try something simpler (and sandbox-friendly) like

```groovy
try {
  sleep 999
} catch (x) {
  echo "aborting soon after $x"
  sleep 10
}
```

?

Reply:
Not 100% sure. I think that would have worked as expected prior to the bug being fixed, but now that it is fixed, https://github.com/jenkinsci/jenkins/blob/435bb7989126d6ef3e3d78ddc5597354b71bff74/core/src/main/java/jenkins/model/queue/ItemDeletion.java#L274 will be called repeatedly and so the second sleep will be aborted. Maybe that's ok though, since it means the test will complete more quickly. I'll test it.

Reply:
Hmm, not sure why that is called repeatedly.

Note that a single call to https://javadoc.jenkins.io/hudson/model/Executor.html#doStop() (for example, clicking red **X**) currently only sends a single interruption to the program, and `CpsFlowExecution` does not redeliver this in case cleanup drags on. Perhaps `workflow-cps` should automatically retry until the build actually aborts? Or should `workflow-job` automatically escalate to `doTerm` and `doKill`?

Reply:
> not sure why that is called repeatedly.

Yes, it is not clear to me either. It was introduced in https://github.com/jenkinsci/jenkins/pull/2789/commits/047e849d09a914ec0178d98e6f6be681e0b213e5, so I will try to take a closer look at that test, but it seems unrelated to Pipeline.

> Perhaps workflow-cps should automatically retry until the build actually aborts? Or should workflow-job automatically escalate to doTerm and doKill?

I also thought about this and wasn't sure. My first instinct is to just fix the immediate issue, and then if we start seeing problems where jobs can't be deleted because builds are not completing quickly enough, we can consider changes like that.

Reply:
One other non-optimal thing about repeated interruptions with the fix is that you can get a _lot_ of [these tasks](https://github.com/jenkinsci/workflow-job-plugin/blob/d82e9f82c8115f04fa3f27fc38e18ca01c6410d2/src/main/java/org/jenkinsci/plugins/workflow/job/WorkflowRun.java#L385-L403) if interruption is not immediate, and I could see it causing system log/build log spam. I didn't have a `BuildWatcher` active previously though, so let me check it again to see if that message gets printed a bunch.

Reply:
> let me check it again to see if that message gets printed a bunch.

The 15 second timer in `printLater` makes that not a problem. From some logging though I can see that around 200 timer tasks get submitted by calls to `interrupt` for a build that takes 10 seconds to stop. I will look at why that `interruptOnDelete` test required repeated calls to `interrupt` because it seems like we don't really want to repeatedly call interrupt like this.

Reply:
Hmm, well `JobTest.interruptOnDelete` passes even if the extra interruptions are removed. Now I am leaning towards removing them, even just for the sake of consistency with other interruption mechanisms in Jenkins.

Reply:
As far as the original question though, yes we can simplify the Pipeline here, but we need two back-to-back calls to `sleep` in the catch block for the issue to reproduce consistently without the fix.

Reply:
And use https://github.com/jenkinsci/jenkins/pull/9790#discussion_r1777792002 for further discussion about repeated interruption.

Reply:
> As far as the original question though, yes we can simplify the Pipeline here, but we need two back-to-back calls to sleep in the catch block for the issue to reproduce consistently without the fix.

In practice actually it seems unnecessary. [3426a6a](https://github.com/jenkinsci/workflow-job-plugin/pull/468/commits/3426a6ac25629e3dba064468f9e3cac9280b0a0c) simplifies the test.